pulp_docker/app/tasks/sync_stages.py

```diff
@@ -0,0 +1,405 @@
+
+        with open(list_downloader.path) as tags_raw:
+            tags_dict = json.loads(tags_raw.read())
+            tag_list = tags_dict['tags']
+        pb.increment()
+        # TODO add pagination
+        log.warn("Tag list len {num}".format(num=len(tag_list)))
+        for tag_name in tag_list:
+            tag_dc = self.create_pending_tag(tag_name)
+            await out_q.put(tag_dc)
+            pb.increment()
+
+        await out_q.put(None)
+
+    def create_pending_tag(self, tag_name):
+        """
+        Create `DeclarativeContent` for each tag.
+
+        Each dc contains enough information to be dowloaded by an ArtifactDownload Stage.
+
+        Args:
+            tag_name (str): Name of each tag
+
+        Returns:
+            pulpcore.plugin.stages.DeclarativeContent: A Tag DeclarativeContent object
+
+        """
+        relative_url = '/v2/{name}/manifests/{tag}'.format(
+            name=self.remote.namespaced_upstream_name,
+            tag=tag_name,
+        )
+        url = urljoin(self.remote.url, relative_url)
+        tag = Tag(name=tag_name)
+        # TODO this one, we don't have anything to add here, so we can't do validation
+        manifest_artifact = Artifact()
+        da = DeclarativeArtifact(
+            artifact=manifest_artifact,
+            url=url,
+            relative_path="TODO-where-should-this-go-{name}".format(name=tag_name),
+            remote=self.remote,
+            # TODO is this necessary for tag list?
+            extra_data={'headers': V2_ACCEPT_HEADERS}
+        )
+        tag_dc = DeclarativeContent(content=tag, d_artifacts=[da])
+        return tag_dc
+
+
+class ProcessContentStage(Stage):
+    """
+    Process all Manifests, Manifest Lists, and Tags.
+
+    For each processed type, create content from nested fields. This stage does not process
+    ManifestBlobs, which do not contain nested content.
+    """
+
+    def __init__(self, remote):
+        """
+        Inform the stage about the remote to use.
+        """
+        self.remote = remote
+
+    async def __call__(self, in_q, out_q):
+        """
+        Create new Content for all unsaved content units with downloaded artifacts.
+
+        Args:
+            in_q(asyncio.Queue): Queue of pulpcore.plugin.stages.DeclarativeContent objects to be
+                                 processed.
+            out_q(asyncio.Queue): Queue of pulpcore.plugin.stages.DeclarativeContent objects that
+                                  have either been processed or were created in this stage.
+
+        """
+        _skipped_schema_1_count = 0
+        _tagged_manifest_list_count = 0
+        _tagged_manifest_count = 0
+        _manifest_list_count = 0
+        _manifest_count = 0
+        while True:
+            dc = await in_q.get()
+            if dc is None:
+                break
+            # If content has been saved, it has also been processed.
+            elif dc.content.pk is not None:
```

Comment:
The `Tag` could have been created by a previous sync that got interrupted (Eg: canceled) after creating the `Tag` but before processing it.  This would result in subsequent syncs never processing the `Tag`, right?

Reply:
All content is processed before it is saved. So in that case, the ManifestList or ImageManifest objects would have been created but could be cancelled before they are saved. So, yeah, that could be a problem. I'll have to think about this. We don't want to process a unit that has already been processed, and whether it was saved seemed convinient. We probably could add something in the dc.extra_data to indicate the current state of each unit.

Reply:
Added "processed" to dc.extra_data to use instead.